src/main/java/edu/wpi/grip/core/Connection.java

```diff
@@ -79,4 +104,13 @@ public void removeConnection(SourceRemovedEvent e) {
             }
         }
     }
+
+    /**
+     * Thrown when the user tries to create a connection that will result in an infinite loop.
+     */
+    public static class InfiniteLoopException extends Error {
```

Comment:
I don't see how that's relevant.

I'd much rather users couldn't use the UI to do an invalid action and cause an exception whenever possible.  I think it's confusing if you *can* connect steps in a loop, but nothing happens and an error is printed to the console or something.

We could also only allow connections form left to right in the pipeline, so the step with the output socket must be before the step with the input socket.  This would also prevent loops, and we wouldn't really lose any functionality by adding that rule (we should add that rule eventually anyways)